sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcDeserializer.scala

```diff
@@ -0,0 +1,216 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution.datasources.orc
+
+import scala.collection.JavaConverters._
+import scala.collection.mutable.ArrayBuffer
+
+import org.apache.hadoop.io._
+import org.apache.orc.mapred.{OrcList, OrcMap, OrcStruct, OrcTimestamp}
+import org.apache.orc.storage.serde2.io.{DateWritable, HiveDecimalWritable}
+
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.expressions.SpecificInternalRow
+import org.apache.spark.sql.catalyst.util._
+import org.apache.spark.sql.execution.datasources.orc.OrcUtils.withNullSafe
+import org.apache.spark.sql.types._
+import org.apache.spark.unsafe.types.UTF8String
+
+private[orc] class OrcDeserializer(
+    dataSchema: StructType,
+    requiredSchema: StructType,
+    missingColumnNames: Seq[String]) {
+
+  private[this] val mutableRow = new SpecificInternalRow(requiredSchema.map(_.dataType))
+
+  private[this] val length = requiredSchema.length
+
+  private[this] val unwrappers = requiredSchema.map(_.dataType).map(unwrapperFor).toArray
+
+  def deserialize(orcStruct: OrcStruct): InternalRow = {
+    var i = 0
+    val names = orcStruct.getSchema.getFieldNames
+    while (i < length) {
+      val name = requiredSchema(i).name
+      val writable = if (missingColumnNames.contains(name)) {
+        null
+      } else {
+        if (names.contains(name)) {
```

Comment:
Doing this kind of thing in the loop is pretty slow, especially on this critical path. Let's pre-compute it before the loop. Again, PLEASE follow the previous code, which did this check when creating the unwrapper.

Reply:
BTW do we really need to handle missing columns for nested fields?

Reply:
1. I moved out the `column-name` logic from `while`, but it still requires `OrcStruct` because we don't have `StructObjectInspector`. So, we cannot move out this from `iter.map`.
2. This is about top-level fields. Could you tell me what do you mean by `nested fields` here?